Draft: Redo "modularization" and "distribution". (#96)
isaqb-org/curriculum-flex: docs/02-modularization/02-learning-goals.adoc

// tag::DE[]

[[LZ-2-1]]

> ==== LZ 2-1: Dekomposition in Bausteine anhand der Anforderungen entwerfen

Reviewer: Not sure why this has been removed?

Author: Editing mistake, fixed in 5c2a70bf0a3984f345fa24daf6a1147a1fbc3944

// tag::DE[]

[[LZ-2-1]]
==== LZ 2-1: Ziele von Modularisierung

> * Architekten verstehen, dass Flexibilität im Kontext dieses Curriculums die Fähigkeit bezeichnet, das System an sich ändernde Anforderungen anzupassen.

Reviewer 2: „Anpassen zu können“, I‘d say

Author: Done in 1ff13c3b70e3ddd2dd85a8e926a5812c6c246e24

> * Architekturen verstehen die Rolle des Geheimnisprinzips beim Erreichen von Flexibilität, insbesondere dem Verstecken von Entscheidungen, die schwierig zu treffen oder später schwer zu ändern sind.

Reviewer 2: ```suggestion
* Architekt:innen verstehen die Rolle des Geheimnisprinzips beim Erreichen von Flexibilität, insbesondere dem Verstecken von Entscheidungen, die schwierig zu treffen oder später schwer zu ändern sind.
```

Author: Done in 407b2516c7479b04f407c0557550e3c5501b1202

* Architekt:innen können die Prinzipien und Regel der Modularisierungsstrategie dokumentieren.

* Architekt:innen wissen, dass Architekturdokumentation aus entsprechenden Modularisierungsmechanismen generiert werden kann.

// end::DE[]

// tag::EN[]

[[LG-2-1]]
==== LG 2-1: Goals of modularization

* Architects understand that flexibility in the context of this curriculum means the ability to adapt the system to changing requirements.

* Architects understand the role of information hiding to achieve flexibility, specifically the hiding of decisions that are difficult to make or may change later.

* Architects can use modularization to map conceptual building blocks to (technical) modules.

* Architects know the impact on modularization of requirements such as independent deployment, parallelization of work, and replaceability of technologies.

[[LG-2-2]]
==== LG 2-2: Coupling

* Architects understand that the goal of modularization is to reduce coupling between modules.

* Architects understand possible causes of tight coupling between modules, such as misplaced module boundaries, ill-designed APIs, and over-constrained communication between modules.

> * Architects understand the special role of mutable state in tight coupling.

Reviewer: This is not present in the German version.

Author: Fixed in 18f22d624670ad7a2ae997c0099796f4f0f9cee2

* Architects understand the role of information hiding to achieve flexibility, specifically the hiding of decisions that are difficult to make or may change later.

* Architects can use modularization to map conceptual building blocks to (technical) modules.

* Architects know the impact on modularization of requirements such as independent deployment, parallelization of work, and replaceability of technologies.

[[LG-2-2]]
==== LG 2-2: Coupling

* Architects understand that the goal of modularization is to reduce coupling between modules.

* Architects understand possible causes of tight coupling between modules, such as misplaced module boundaries, ill-designed APIs, and over-constrained communication between modules.

* Architects understand the special role of mutable state in tight coupling.

* Architects can design a modularization in a bottom-up fashion that will keep coupling low.

* Architects can decrease coupling by increasing cohesion within modules.

[[LG-2-3]]
==== LG 2-3: Modularization and the organization

* Architects can analyze how module boundary choices affect complexity, coordination effort for changes, and development and maintenance costs.

> * Architects understand the role of Conway's law in constraining modularization.

Reviewer: Add Inverse Conway?

Author: Good point, added in 12bec11b99b4d56980d3f236ab4907ebfb896bf7

* Architects can identify organizational and technical costs for integrating a modular system.

* Architects understand the different communication challenges in organizations that organize teams by bounded contexts and those that organize them by technical competency.

[[LG-2-4]]
==== LG 2-4: Module mechanisms

* Architects understand programming-language mechanisms that relate to modularization such as classes and interfaces in object-oriented languages, namespace mechanisms, and ML-style module languages.

* Architects understand the role of opaque interfaces in enforcing loose coupling.

* Architects know interface languages and their differences such as Java/C# interfaces, Java modules, and ML-style module signatures, and can leverage them to enforce loose coupling.

* Architects know non-linguistic (not built into the programming language) approaches to enforcing modularization such as creating unit tests and using external tools that operate on the module dependency graph.

* Architects understand the relationship between module mechanisms and deployment strategies.

[[LG-2-5]]
==== LG 2-5: Modularization documentation

> * Architects understand that architecture documentation and implementation modularization should be aligned.

Reviewer 2: ```suggestion
* Architects understand that architecture documentation and implementation modularization must be aligned.
```

Reviewer 2: "Resolved" without comment?

> * Architects can document the principles and rules of the modularization strategy.

Reviewer 2: ```suggestion
* Architects are able to document the principles and rules of the modularization strategy.
```

Author: Why?
